enarx-keep-sev-shim/src/addr.rs

```diff
@@ -0,0 +1,65 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Some basic address operations
+
+/// The offset of shim virtual address space to the physical address
+///
+/// physical address + `SHIM_VIRT_OFFSET` = shim virtual address
+///
+/// FIXME: change to dynamic offset with ASLR
+const SHIM_VIRT_OFFSET: u64 = 0xFFFF_FF80_0000_0000;
+
+use core::convert::TryFrom;
+use memory::Address;
+
+/// Address in the shim physical address space
+pub struct ShimPhysAddr<U>(Address<u64, U>);
+
+impl<T, U> From<ShimPhysAddr<U>> for memory::Register<T>
+where
+    memory::Register<T>: From<memory::Register<u64>>,
+{
+    #[inline(always)]
+    fn from(val: ShimPhysAddr<U>) -> Self {
+        memory::Register::<u64>::from(val.0).into()
+    }
+}
+
+/// Address in the shim virtual address space
+pub struct ShimVirtAddr<U>(Address<u64, U>);
+
+impl<U> ShimVirtAddr<U> {
+    /// Converts a `ShimVirtAddr` to an `Address`
+    #[inline(always)]
+    pub(crate) fn into_address<T>(self) -> Address<T, U>
+    where
+        Address<T, U>: From<Address<u64, U>>,
+    {
+        self.0.into()
+    }
+
+    /// Converts a `ShimVirtAddr` to a physical `ShimPhysAddr`
+    #[inline(always)]
+    pub(crate) fn into_phys_address(self) -> ShimPhysAddr<U> {
+        // Safety: checked, that it is in the shim virtual address space earlier
+        #[allow(clippy::integer_arithmetic)]
+        ShimPhysAddr(unsafe {
+            Address::unchecked(self.into_address::<u64>().raw() - SHIM_VIRT_OFFSET)
+        })
+    }
+}
```

Comment:
Why not `impl From` for these types instead of `into_address` or `into_phys_address`?

Reply:
With:
```rust
impl<T, U> From<ShimVirtAddr<U>> for Address<T, U>
where
    Address<T, U>: From<Address<u64, U>>,
{
    fn from(shim_virt_addr: ShimVirtAddr<U>) -> Self {
        shim_virt_addr.into_address().into()
    }
}
```
we get:
```
error[E0119]: conflicting implementations of trait `core::convert::From<addr::ShimVirtAddr<()>>` for type `memory::address::Address<addr::ShimVirtAddr<()>, ()>`:
  --> src/addr.rs:68:1
   |
68 | / impl<T, U> From<ShimVirtAddr<U>> for Address<T, U>
69 | | where
70 | |     Address<T, U>: From<Address<u64, U>>,
71 | | {
...  |
74 | |     }
75 | | }
   | |_^
   |
   = note: conflicting implementation in crate `memory`:
           - impl<T> core::convert::From<T> for memory::address::Address<T, ()>;
   = note: upstream crates may add a new impl of trait `core::convert::From<memory::address::Address<u64, ()>>` for type `memory::address::Address<addr::ShimVirtAddr<()>, ()>` in future versions
```

and with:
```rust
mpl<U> From<ShimVirtAddr<U>> for Address<u64, U> {
    fn from(shim_virt_addr: ShimVirtAddr<U>) -> Self {
        shim_virt_addr.into_address()
    }
}
```

the relevant code changes look ugly:
```diff
@@ -54,8 +54,12 @@ impl HostCall {
     /// Panics, if it is initialized more than once.
     #[inline(always)]
     pub fn init(address: ShimVirtAddr<*mut Page>) {
-        let prev =
-            unsafe { HOSTCALL_VIRT_ADDR.swap(address.into_address().raw(), Ordering::Release) };
+        let prev = unsafe {
+            HOSTCALL_VIRT_ADDR.swap(
+                Address::<usize, _>::from(Address::<u64, _>::from(address)).raw(),
+                Ordering::Release,
+            )
+        };
         if prev != 0 {
             panic!("HostCall already initialized");
         }
```

Reply:
```rust
impl<T, U> From<ShimVirtAddr<U>> for Address<T, U>
where
    Address<T, U>: From<Address<u64, U>>,
{
    fn from(shim_virt_addr: ShimVirtAddr<U>) -> Self {
        shim_virt_addr.into_address().into()
    }
}
```
would need to look like:
```rust
impl<T, U> From<ShimVirtAddr<U>> for Address<T, U>
where
    Address<T, U>: From<Address<u64, U>>,
    T : !ShimVirtAddr
{
    fn from(shim_virt_addr: ShimVirtAddr<U>) -> Self {
        shim_virt_addr.into_address().into()
    }
}
```

Reply:
Anyway, I simplified it by converting
```diff
-static mut HOSTCALL_VIRT_ADDR: AtomicUsize = AtomicUsize::new(0);
+static mut HOSTCALL_VIRT_ADDR: AtomicU64 = AtomicU64::new(0);
```